spark/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala

```diff
@@ -238,6 +250,90 @@ private[delta] class ConflictChecker(
+   * materializes and recommits all existing files in table using one or several commits to ensure
+   * that every AddFile has a base row ID and a default row commit version. When enabling
+   * row tracking on an existing table, the following occurs:
+   *    1. (If necessary) Protocol upgrade + Table Feature Support is added
+   *    2. RowTrackingBackfill commit(s)
+   *    3. Table property and metadata are updated.
+   * RowTrackingBackfill does not do any data change. It doesn't matter whether a file is
+   * recommitted after the table feature support from Backfill or some other concurrent transaction;
+   * every AddFile just needs to have a base row ID and a default row commit version somehow.
+   * However, correctness issues can arise if we don't do the checks in this method.
+   *
+   * Check that RowTrackingBackfill is not resurrecting files that were removed concurrently and
+   * that an AddFile and its corresponding RemoveFile have the same base row ID and
+   * default row commit version. To do this, we:
+   *    1. remove AddFile's from a backfill commit if an AddFile or a RemoveFile with the same path
+   *       was added in the winning concurrent transactions. Files in a winning transaction can be
+   *       removed from backfill because they were already re-committed.
+   *    2. copy over base row IDs and default row commit versions if the current transaction re-adds
+   *       or delete an AddFile with the same path as an Addfile from a winning backfill commit.
+   */
+  private def resolveRowTrackingBackfillConflicts(): Unit = {
+    // If row tracking is not supported, there can be no backfill commit.
+    if (!RowTracking.isSupported(currentTransactionInfo.protocol)) {
+      assert(!currentTransactionInfo.isRowTrackingBackfillTxn)
+      assert(!winningCommitSummary.isRowTrackingBackfillTxn)
+      return
+    }
+
+    val timerPhaseName = "checked-row-tracking-backfill"
+    if (currentTransactionInfo.isRowTrackingBackfillTxn) {
+      recordTime(timerPhaseName) {
+        // Any winning commit seen by backfill must have row IDs and row commit versions, because
+        // `reassignOverlappingRowIds` will add a base row ID and `reassignRowCommitVersions`
+        // will add a default row commit versions to all files. So we don't need
+        // Backfill to commit the same file again.
+        val filePathsToRemoveFromBackfill = winningCommitSummary.actions.collect {
+          case a: AddFile => a.path
+          case r: RemoveFile => r.path
+        }.toSet
+
+        // Remove files from this Backfill commit if they were removed or re-committed by
+        // a concurrent winning txn.
+        if (filePathsToRemoveFromBackfill.nonEmpty) {
+          val newActions = currentTransactionInfo.actions.filterNot {
+            case a: AddFile => filePathsToRemoveFromBackfill.contains(a.path)
+            case d: DomainMetadata if RowTrackingMetadataDomain.isSameDomain(d) => false
```

Comment:
We keep the row tracking metadata action but it's not accurate anymore though - the winning transaction probably increased the high-water mark. I assume we run the conflict resolution rule to re-assign the high-water mark afterwards?

Might be worth calling out that ordering dependency in a comment where these rules are called

Reply:
@johanl-db Yes, this will be handled by the reassignOverlappingRowIds() that we on purpose place after the resolveRowTrackingBackfillConflicts() in checkConflicts(), adding a comment here for extra clarity.